taskcluster/app_services_taskgraph/transforms/beetmover.py

```diff
@@ -2,74 +2,210 @@
-        version = get_version(config.params)
-        publications = module_info["publications"]
+        for key in ("worker.action",):
+            resolve_keyed_by(
+                task, key, item_name=task["name"], level=config.params["level"]
+            )
+        yield task
+
+
+def _build_upstream_artifacts(config, task):
+    upstream_artifacts = []
+
+    for label, dep in config.kind_dependencies_tasks.items():
+        if label not in task["dependencies"].values():
+            continue
+
+        paths = sorted(
+            artifact["name"] for artifact in dep.attributes.get("release-artifacts", [])
+        )
+
+        if paths:
+            upstream_artifacts.extend(
+                [
+                    {
+                        "taskId": {"task-reference": f"<{dep.kind}>"},
+                        "taskType": "repackage",
+                        "paths": paths,
+                    }
+                ]
+            )
 
-        worker_definition = {"upstream-artifacts": [{
+    return upstream_artifacts
+
+
+def _build_maven_upstream_artifacts(config, task):
+    module_info = task["attributes"]["buildconfig"]
+    version = get_version(config.params)
+    publications = module_info["publications"]
+
+    return [
+        {
             "taskId": {"task-reference": "<module-build>"},
             "taskType": "build",
-            "paths": (publications_to_artifact_paths(version, publications,
-                                                     ("", ".sha1", ".md5"))),
-        }, {
+            "paths": (
+                publications_to_artifact_paths(
+                    version, publications, ("", ".sha1", ".md5")
+                )
+            ),
+        },
+        {
             "taskId": {"task-reference": "<signing>"},
             "taskType": "signing",
             "paths": (publications_to_artifact_paths(version, publications, (".asc",))),
-        }]}
+        },
+    ]
 
+
+@transforms.add
+def build_upstream_artifacts(config, tasks):
+    for task in tasks:
         task.setdefault("worker", {})
-        task["worker"].update(worker_definition)
+        if task["worker"].get("action") == "push-to-maven":
+            task["worker"]["upstream-artifacts"] = _build_maven_upstream_artifacts(
+                config, task
+            )
+
+        else:
+            task["worker"]["upstream-artifacts"] = _build_upstream_artifacts(
+                config, task
+            )
+
         yield task
 
 
 @transforms.add
-def build_artifact_map(config, tasks):
+def split_by_release_phase(config, tasks):
     for task in tasks:
-        module_info = task["attributes"]["buildconfig"]
-        version = get_version(config.params)
+        attributes = task.setdefault("attributes", {})
+        if attributes.get("shipping_phase"):
+            yield task
+            continue
+
+        for phase in config.graph_config["release-promotion"]["flavors"]:
+            reltask = deepcopy(task)
```

Comment:
@ahal I'm sure you know better than I, what is the guidance on using deepcopy vs copy_task?

Reply:
I guess I don't because at first I didn't know what `copy_task` was! But looks like it only exists in `gecko_taskgraph`, so it's not an option here. Looks like `copy_task` claims to be more efficient than `deepcopy`.. though I wonder how true that actually is. It also supports copying `Task` objects, but that's not really necessary in transforms.

I'd be open to upstreaming it, but it might be worth doing some performance tests to validate whether it's actually faster or not.